src/main/java/uk/gov/homeoffice/digital/sas/timecard/validators/timeentry/TimeEntryValidator.java

```diff
@@ -30,6 +35,102 @@ public boolean isValid(Object value, ConstraintValidatorContext context) {
+      var payload = timeEntryClashes.stream().map(this::transformTimeEntry)
+          .collect(Collectors.toCollection(ArrayList::new));
+
+      HibernateConstraintValidatorContext hibernateContext =
+          context.unwrap(HibernateConstraintValidatorContext.class);
+
+      hibernateContext.disableDefaultConstraintViolation();
+      hibernateContext.withDynamicPayload(payload);
+      hibernateContext
+          .buildConstraintViolationWithTemplate(
+              "Time periods must not overlap with another time period")
+          .addPropertyNode(clashingProperty.toString())
+          .addConstraintViolation();
+      return false;
+    }
+    return true;
+  }
+
+  private JSONObject transformTimeEntry(TimeEntry timeEntry) {
+    var result = new JSONObject();
+    result.put("startTime", timeEntry.getActualStartTime());
+    result.put("endTime", timeEntry.getActualEndTime());
+    result.put("timePeriodTypeId", timeEntry.getTimePeriodTypeId());
+    return result;
+  }
+
+  private ClashingProperty getClashingProperty(
+      TimeEntry timeEntry, List<TimeEntry> timeEntryClashes) {
+    var startTimeClash = false;
+    var endTimeClash = false;
+
+    for (TimeEntry timeEntryClash : timeEntryClashes) {
+      if (timeEntry.getActualStartTime().equals(timeEntryClash.getActualStartTime())) {
+        startTimeClash = true;
+      }
+
+      if (timeEntryClash.getActualEndTime() != null
+          && startTimeClashesWithTimeEntry(timeEntry.getActualStartTime(), timeEntryClash)) {
+        startTimeClash = true;
+      }
+
+      if (timeEntry.getActualEndTime() != null
+          && timeEntryClash.getActualEndTime() != null
+          && endTimeClashesWithTimeEntry(timeEntry.getActualEndTime(), timeEntryClash)) {
+        endTimeClash = true;
+      }
+
+      if (timeEntry.getActualEndTime() != null
+          && timeEntryClash.getActualEndTime() != null
+          && startTimeClashesWithTimeEntry(timeEntryClash.getActualStartTime(), timeEntry)
+      ) {
+        endTimeClash = true;
+      }
+
+      if (startTimeClash && endTimeClash) {
+        break;
+      }
+    }
+
+    if (startTimeClash && endTimeClash) {
+      return ClashingProperty.START_AND_END_TIME;
+    }
+    if (startTimeClash) {
+      return ClashingProperty.START_TIME;
+    }
+    if (endTimeClash) {
+      return ClashingProperty.END_TIME;
+    }
+
+    return ClashingProperty.START_AND_END_TIME;
+  }
+
+  private boolean endTimeClashesWithTimeEntry(Date endTime, TimeEntry timeEntry) {
+    return (endTime.after(timeEntry.getActualStartTime())
+        && endTime.getTime() <= timeEntry.getActualEndTime().getTime());
+  }
+
+  private boolean startTimeClashesWithTimeEntry(Date startTime, TimeEntry timeEntry) {
+    return (startTime.getTime() >= timeEntry.getActualStartTime().getTime()
+        && startTime.before(timeEntry.getActualEndTime()));
+  }
+
+  enum ClashingProperty {
```

Comment:
Can we extract this outside of this enum outside of this class? Potentially into a separate enums package.